### PR TITLE
ROC-4776: Transforming claim into ccd claim

### PR DESCRIPTION
**Pull request by dharmendrak**

### CODE REVIEW

File: ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/deserialize/AmountBreakDownMixIn.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import java.io.IOException;
+import java.util.List;
+import java.util.ListIterator;
+
+public abstract class AmountBreakDownMixIn {
+
+    public AmountBreakDownMixIn() {
+    }
+
+    public class AmountRowSerializer extends StdSerializer<List> {
```
Comment (srineelesh) [inline ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/deserialize/AmountBreakDownMixIn.java]: better to extend from JsonSerializer

File: ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/deserialize/CompanyDetailsMixIn.java [hunk 2]
```diff
@@ -0,0 +1,6 @@
+package uk.gov.hmcts.cmc.ccd.deserialize;
```
Comment (srineelesh) [inline ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/deserialize/CompanyDetailsMixIn.java]: this should not be in deserialize package

File: ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/deserialize/ClaimDataMixIn.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+import uk.gov.hmcts.cmc.domain.models.otherparty.TheirDetails;
+import uk.gov.hmcts.cmc.domain.models.party.Party;
+
+import java.math.BigDecimal;
+import java.util.List;
+import java.util.Optional;
+
+public abstract class ClaimDataMixIn {
+
+    @JsonProperty("listOfClaimants")
```
Comment (sabahirfan) [inline ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/deserialize/ClaimDataMixIn.java]: Can we have just `claimants` and `defendants`?  Other collections does not have prefix like `listOf...`.

File: ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/deserialize/InterestBreakDownMixIn.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+import uk.gov.hmcts.cmc.domain.models.InterestDate;
+
+import java.math.BigDecimal;
+
+public abstract class InterestBreakDownMixIn {
+
+    @JsonProperty("interestBreakDownAmount")
+    abstract BigDecimal getTotalAmount();
+
+    @JsonProperty("interestBreakDownCalc")
```
Comment (sabahirfan) [inline ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/deserialize/InterestBreakDownMixIn.java]: IMHO `interestBreakDownExplanation` would be more consistent.

File: ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/deserialize/InterestMixIn.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+import java.math.BigDecimal;
+import java.time.LocalDate;
+
+public abstract class InterestMixIn {
+
+    @JsonProperty("interestType")
+    abstract InterestType getType();
+
+    @JsonProperty("InterestRate")
```
Comment (sabahirfan) [inline ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/deserialize/InterestMixIn.java]: `interestRate` with small `i` please

File: ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/deserialize/OrganisationMixIn.java [hunk 6]
```diff
@@ -0,0 +1,12 @@
+package uk.gov.hmcts.cmc.ccd.deserialize;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public abstract class OrganisationMixIn extends PartyMixIn {
+
+    @JsonProperty("partyContactPerson")
```
Comment (sabahirfan) [inline ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/deserialize/OrganisationMixIn.java]: IMHO `organisationContactPerson` and `organisationCompaniesHouseNumber` are more meaningful

File: ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/deserialize/RepresentativeMixIn.java [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+public abstract class RepresentativeMixIn {
+
+    @JsonProperty("representativeOrganisationName")
+    abstract String getOrganisationName();
+
+    @JsonProperty("representativeOrganisationAddress")
+    abstract Address getOrganisationAddress();
+
+    @JsonUnwrapped(prefix = "representativeOrganisation")
```
Comment (sabahirfan) [inline ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/deserialize/RepresentativeMixIn.java]: 👍

File: src/main/java/uk/gov/hmcts/cmc/claimstore/config/JacksonConfiguration.java [hunk 8]
```diff
@@ -25,4 +56,29 @@ public ObjectMapper objectMapper() {
             .disable(SerializationFeature.FAIL_ON_EMPTY_BEANS);
     }
 
+    @Bean
+    public ObjectMapper ccdObjectMapper() {
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/config/JacksonConfiguration.java]: Would be nice to this configuration in ccd-adapter module under `uk.gov.hmcts.cmc.ccd.config`

File: ccd-adapter/src/main/java/uk/gov/hmcts/cmc/custom/deserializer/ListItemDeserializer.java [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class ListItemDeserializer extends JsonDeserializer<List<JsonNode>> {
+
+    @Override
+    public List<JsonNode> deserialize(JsonParser jsonParser, DeserializationContext deserializationContext) throws IOException, JsonProcessingException {
+        JsonNode node = jsonParser.getCodec().readTree(jsonParser);
+        ArrayList<JsonNode> retunList = new ArrayList<>();
```
Comment (sabahirfan) [inline ccd-adapter/src/main/java/uk/gov/hmcts/cmc/custom/deserializer/ListItemDeserializer.java]: typo in `retunList`

File: src/test/java/uk/gov/hmcts/cmc/claimstore/processors/MixMapperTest.java [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+import uk.gov.hmcts.cmc.domain.models.otherparty.TheirDetails;
+import uk.gov.hmcts.cmc.domain.models.party.Organisation;
+import uk.gov.hmcts.cmc.domain.models.party.Party;
+import uk.gov.hmcts.cmc.domain.models.sampledata.SampleClaimData;
+import uk.gov.hmcts.cmc.domain.models.sampledata.SampleParty;
+import uk.gov.hmcts.cmc.domain.models.sampledata.SampleTheirDetails;
+
+import static uk.gov.hmcts.cmc.ccd.assertion.Assertions.assertThat;
+
+public class MixMapperTest {
```
Comment (sabahirfan) [inline src/test/java/uk/gov/hmcts/cmc/claimstore/processors/MixMapperTest.java]: Please move the tests in ccd-adapter modules as well.

File: src/test/java/uk/gov/hmcts/cmc/claimstore/processors/MixMapperTest.java [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+
+    private JsonMapper processor = new JsonMapper(new JacksonConfiguration().ccdObjectMapper());
+
+    @Test
+    public void shouldProcessIndividualToCCDJson() throws JSONException {
+        //given
+        Party individual = SampleParty.builder().individual();
+
+        System.out.println(processor.toJson(individual));
```
Comment (sabahirfan) [inline src/test/java/uk/gov/hmcts/cmc/claimstore/processors/MixMapperTest.java]: Don't forget to remove these `System.out.println(..)`.

File: ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/deserialize/PartyMixIn.java [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+    @JsonProperty("partyDateOfBirth")
+    abstract LocalDate getDateOfBirth();
+
+    @JsonProperty("partyName")
+    abstract String getName();
+
+    @JsonProperty("partyAddress")
+    abstract Address getAddress();
+
+    @JsonProperty("partyEmail")
```
Comment (ckalista) [inline ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/deserialize/PartyMixIn.java]: Where does that email come from? Can't see it in Party...

File: domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/party/PartyType.java [hunk 13]
```diff
@@ -0,0 +1,8 @@
+package uk.gov.hmcts.cmc.domain.models.party;
+
+public enum PartyType {
```
Comment (sabahirfan) [inline domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/party/PartyType.java]: Do we need this enum?

File: ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/jackson/mixin/PartyMixIn.java [hunk 14]
```diff
@@ -0,0 +1,34 @@
+package uk.gov.hmcts.cmc.ccd.jackson.mixin;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonUnwrapped;
+import uk.gov.hmcts.cmc.domain.models.Address;
+import uk.gov.hmcts.cmc.domain.models.legalrep.Representative;
+
+import java.time.LocalDate;
+
+public abstract class PartyMixIn {
```
Comment (ckalista) [inline ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/jackson/mixin/PartyMixIn.java]: When party is serialised in defendant list it should be prefixed `claimantProvidedParty` and not `party`.

Comment (ckalista) [inline ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/jackson/mixin/PartyMixIn.java]: We are adding a lot of customisation just to prefix every attributes with `party`. May be worth investigation if we can so more Jackon magic here?

File: ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/jackson/custom/deserializer/ListItemDeserializer.java [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+import com.fasterxml.jackson.databind.JsonNode;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class ListItemDeserializer extends JsonDeserializer<List<JsonNode>> {
+
+    @Override
+    public List<JsonNode> deserialize(JsonParser jsonParser, DeserializationContext deserializationContext) throws IOException, JsonProcessingException {
```
Comment (sabahirfan) [inline ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/jackson/custom/deserializer/ListItemDeserializer.java]: `JsonProcessingException` is not required in throws clause

File: src/test/java/uk/gov/hmcts/cmc/claimstore/processors/MixMapperTest.java (hunk 10 above)
Comment (sabahirfan) [reply src/test/java/uk/gov/hmcts/cmc/claimstore/processors/MixMapperTest.java]: I don't see that. This is all you need. 
```
    private JsonMapper processor = new JsonMapper(new CCDAdapterConfig().ccdObjectMapper());
```

File: src/test/java/uk/gov/hmcts/cmc/claimstore/processors/MixMapperTest.java [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+        String json = processor.toJson(claimData);
+
+        ClaimData output = processor.fromJson(json, ClaimData.class);
+        String outputJson = processor.toJson(output);
+
+        assertThat(json, CoreMatchers.equalTo(outputJson));
+    }
+
+    @Test
+    public void shouldProcessAmountBreakDownToCCDJson() throws JSONException {
```
Comment (sabahirfan) [inline src/test/java/uk/gov/hmcts/cmc/claimstore/processors/MixMapperTest.java]: Please have separate unit test classes, currently all of these are in this one class.

File: src/test/java/uk/gov/hmcts/cmc/claimstore/processors/MixMapperTest.java (hunk 10 above)
Comment (sabahirfan) [reply src/test/java/uk/gov/hmcts/cmc/claimstore/processors/MixMapperTest.java]: Probably time to move `JsonMapper` in common place. We should have the tests in same module where we have code.

File: ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/jackson/mixin/RepresentativeMixIn.java [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+@SuppressWarnings("squid:S1610")
+public abstract class RepresentativeMixIn {
+
+    @JsonProperty("representativeOrganisationName")
+    abstract String getOrganisationName();
+
+    @JsonProperty("representativeOrganisationAddress")
+    abstract Address getOrganisationAddress();
+
+    @JsonUnwrapped(prefix = "representativeOrganisation")
```
Comment (sabahirfan) [inline ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/jackson/mixin/RepresentativeMixIn.java]: Is this capitalising correcting when applying prefix? I means is it generating `representativeOrganisationphone` or `representativeOrganisationPhone` ?

File: domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/amount/AmountType.java [hunk 18]
```diff
@@ -0,0 +1,7 @@
+package uk.gov.hmcts.cmc.domain.models.amount;
+
+public enum AmountType {
```
Comment (sabahirfan) [inline domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/amount/AmountType.java]: I think we don't need these extra enums.

**Pull request merged**